### PR TITLE
TVOS-175: Model Validation

### PR DESCRIPTION
**Pull request by huebnerob**
# Ticket(s)

https://vimean.atlassian.net/browse/TVOS-175

# Ticket Summary

DEV | As a developer, I'd like VimeoNetworking to enforce non-optional "resourceKey" and "uri" properties on all model objects

# Implementation Summary

1. Merges VIMObjectMapper into VimeoNetworking as a direct import.
This facilitated making changes that were necessary for implementing
validation.

2. Implements model object validation.  A new method is added to all
instances of VIMModelObject: validateModel.  This gives model objects
an opportunity to throw an error if they deem themselves invalid.  As
our first steps into this process, we have `VIMVideo` checking for
non-nil `uri` and `resourceKey`, and `VIMUser` checking for non-nil
`uri`.  

3. (⭐ **BONUS**⭐ ) Tests were added for these validations as well. cough @jasonhawkins 

# How to test

Either programmatically or with Charles Proxy, generate VIMVideo objects with a nil uri or resourceKey, or VIMUser objects with a nil uri.  These objects should fail validation and return a `.Failure(NSError)` from VimeoNetworking now.

### CODE REVIEW

File: VimeoNetworking/VimeoNetworking.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -14,6 +14,11 @@
 		01B5C8681CC826D800694F4A /* VimeoNetworking.h in Headers */ = {isa = PBXBuildFile; fileRef = 01B5C8671CC826D800694F4A /* VimeoNetworking.h */; settings = {ATTRIBUTES = (Public, ); }; };
 		01B5C86F1CC826D800694F4A /* VimeoNetworking.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 01B5C8641CC826D800694F4A /* VimeoNetworking.framework */; };
 		01B5C8741CC826D800694F4A /* VimeoNetworkingTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 01B5C8731CC826D800694F4A /* VimeoNetworkingTests.swift */; };
+		01CDBEFE1CEBA2030093DDF4 /* VIMMappable.h in Headers */ = {isa = PBXBuildFile; fileRef = 01CDBEF91CEBA2030093DDF4 /* VIMMappable.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
Comment (alfiehanssen) [inline VimeoNetworking/VimeoNetworking.xcodeproj/project.pbxproj]: Can we make this a subspec?

Comment (alfiehanssen) [inline VimeoNetworking/VimeoNetworking.xcodeproj/project.pbxproj]: Can you add a note to the top of the VIMObjectMapper repo pointing people to VimeoNetworking and the subspec?

File: VimeoNetworking/VimeoNetworking/Models/VIMVideo.m [hunk 2]
```diff
@@ -139,6 +139,34 @@ - (void)didFinishMapping
...
+    
+    if (self.uri == nil)
+    {
+        NSString *description = @"VIMVideo failed validation: uri cannot be nil";
+        *error = [NSError errorWithDomain:VIMModelObjectErrorDomain code:VIMModelObjectValidationErrorCode userInfo:@{NSLocalizedDescriptionKey: description}];
+        
+        return;
+    }
+    
+    if (self.resourceKey == nil)
```
Comment (alfiehanssen) [inline VimeoNetworking/VimeoNetworking/Models/VIMVideo.m]: Should the resource key check be above the uri check?

**Comment by alfiehanssen** [comment]
Boo yah, CLASSIC PR, One of the All Time Favorites 

DWR

File: VimeoNetworking/VimeoNetworking/Models/VIMVideo.m (hunk 2 above)
Comment (huebnerob) [inline VimeoNetworking/VimeoNetworking/Models/VIMVideo.m]: tomato tomahto, only matters if both are nil, and I feel like if both are nil we're in deep 💩

File: VimeoNetworking/VimeoNetworking.xcodeproj/project.pbxproj (hunk 1 above)
Comment (huebnerob) [inline VimeoNetworking/VimeoNetworking.xcodeproj/project.pbxproj]: 👍

**Pull request merged**